app/app.go

```diff
@@ -519,35 +524,52 @@ func (app *App) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) abci.R
 func (app *App) EndBlocker(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
 	res := app.mm.EndBlock(ctx, req)
 	// NOTE: this is a specific feature for upgrading from v1 to v2. It will be deprecated in v3
-	if app.UpgradeKeeper.ShouldUpgradeToV2(req.Height) {
-		if app.AppVersion(ctx) == v1.Version {
-			app.SetAppVersion(ctx, v2.Version)
+	if app.UpgradeKeeper.ShouldUpgradeToV2(req.Height) && app.AppVersion(ctx) == v1.Version {
+		if err := app.Upgrade(ctx, v2.Version); err != nil {
+			panic(err)
 		}
 		// from v2 to v3 and onwards we use a signalling mechanism
 	} else if shouldUpgrade, version := app.UpgradeKeeper.ShouldUpgrade(); shouldUpgrade {
 		// Version changes must be increasing. Downgrades are not permitted
 		if version > app.AppVersion(ctx) {
-			app.SetAppVersion(ctx, version)
-			app.UpgradeKeeper.ResetTally(ctx, version)
+			if err := app.Upgrade(ctx, version); err != nil {
+				panic(err)
+			}
 		}
 	}
 	return res
 }
 
+func (app *App) Upgrade(ctx sdk.Context, version uint64) error {
+	app.SetAppVersion(ctx, version)
+	return app.mm.RunMigrations(ctx, app.configurator, app.AppVersion(ctx), version)
+}
+
 // InitChainer application update at chain initialization
 func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.ResponseInitChain {
 	var genesisState GenesisState
 	if err := tmjson.Unmarshal(req.AppStateBytes, &genesisState); err != nil {
 		panic(err)
 	}
-	return app.mm.InitGenesis(ctx, app.appCodec, genesisState)
+	// genesis must always contain the consensus params. The validator set howerver is derived from the
+	// initial genesis state
```

Comment:
I don't understand this comment, can it be rephrased? The consensus params and the validator set seem orthogonal. 

Also the validator set doesn't seem applicable to the following three lines so if it's not relevant in this context, proposal to delete that portion of the comment.

Reply:
I could perhaps move this comment to the function signature. I think it's important since normally there are multiple ways of setting the consensus params and validator set that we clarify which one we use